doc/api18970.md

```diff
@@ -0,0 +1,288 @@
+  function.)
+* A thread is **preempted** when it is *Running* but the scheduler
+  (involuntarily) transitions it to *Ready*.
+
+A **context switch** occurs when the current thread transitions out of
+*Running* (voluntarily or involuntarily) and a different thread is
+selected as the current thread.
+
+**NOTE** The term "sleeping" when applied to Zephyr threads has
+historically meant a thread that is *Waiting* for a timeout event.
+
+**NOTE** Issue 20802 introduces the possibility that a cooperative
+thread may yield when it invokes a function like `k_mutex_release()`
+that results in a context switch.
+
+For API behavior we are primarily interested in whether invoking a
+function can cause a context switch from a cooperative thread.
+
+### Context Terminology
+
+**TODO** fill this out
+
+This section is intended to describe the privilege and processor context
+variations in which a thread can run.  It should provide or reference
+definitions of terms like this:
+
+* Kernel [initialization
+  level](https://docs.zephyrproject.org/latest/reference/drivers/index.html#initialization-levels)
+* User-space versus kernel (system call)
+* Normal ("thread"?) versus interrupt (invoked from an Interrupt Service
+  Routine)
+
+For API behavior we are interested in whether a particular function
+**may**, **must**, or **must not** be invoked from a specific context.
+
+### Function vs Operation
+
+tl;dr: A function *returns* while an operation *completes*.
+
+A **function** is an addressible sequence of actions to which control is
+transferred by **invoking** the function with various parameters, and
+from which control is transferred back to the point where it was invoked
+when the function **returns**.
+
+A function may succeed or fail.  In many cases failure is indicated by a
+negative integer return value while success is indicated by a
+non-negative integer return value, but specific functions may use other
+conventions.
+
+An **operation** is behavior, which is generally initiated by invoking a
+function.  An operation can also succeed or fail.  An operation
+**completes** when the success or failure of the operation has been
+determined and made available through whatever mechanism was specified
+through the initiating function.
+
+For APIs we are particularly interested in the relationship between the
+function and its associated operation.  Some potential relationships
+include:
+* The function succeeds if and only if its associated operation
+  completes with success. (This is usually expected of *synchronous*
+  functions.)
+* If the function fails then its associated operation has not been
+  initiated.  If the function succeeds then its associated operation has
+  been initiated.  The success or failure of the operation is not
+  otherwise coupled to success or failure of its initiating function.
+  (This is usually expected of *asynchronous* functions.)
+
+## Definitions
+
+### suspendable (function)
```

Comment:
We're having trouble finding a word for this that we both like :).

My concern with this latest attempt is that a function which causes the thread to enter the waiting state is suspendable by this definition. This could be confusing, since the reader might thing "suspendable" means "may enter the suspended state".

What about a straightforward "may switch context"?

Reply:
See new update that uses *rescheduling* and refined terms.